diff -u -bur thttpd-2.25b/config.h thttpd-2.25b-patched/config.h
--- thttpd-2.25b/config.h	2005-06-29 10:53:18.000000000 -0700
+++ thttpd-2.25b-patched/config.h	2008-06-26 15:47:56.000000000 -0700
@@ -75,7 +75,7 @@
 ** or whatever.  If you don't want any limit, comment this out, but that's
 ** probably a really bad idea.
 */
-#define CGI_TIMELIMIT 30
+#define CGI_TIMELIMIT 300
 
 /* CONFIGURE: Maximum number of simultaneous CGI programs allowed.
 ** If this many are already running, then attempts to run more will
@@ -276,7 +276,7 @@
 
 /* CONFIGURE: $PATH to use for CGI programs.
 */
-#define CGI_PATH "/usr/local/bin:/usr/ucb:/bin:/usr/bin"
+#define CGI_PATH "/bin:/sbin:/usr/bin:/usr/sbin"
 
 /* CONFIGURE: If defined, $LD_LIBRARY_PATH to use for CGI programs.
 */
